src/jvm/main/org/jetbrains/kotlinx/lincheck/CTestStructure.java

```diff
@@ -110,43 +94,53 @@ private static void readTestStructureFromClass(
                     String nameInOperation = opAnn.params().length > 0 ? opAnn.params()[i] : null;
                     Parameter parameter = m.getParameters()[i];
                     ParameterGenerator<?> parameterGenerator = getOrCreateGenerator(m, parameter, nameInOperation, namedGens, defaultGens, randomProvider);
-                    parameterGeneratorsMap.putIfAbsent(parameter.getType(), parameterGenerator);
+                    testStructureBuilder.parameterGeneratorsMap.putIfAbsent(parameter.getType(), parameterGenerator);
                     gens.add(parameterGenerator);
                 }
                 ActorGenerator actorGenerator = new ActorGenerator(m, gens, opAnn.runOnce(),
                     opAnn.cancellableOnSuspension(), opAnn.allowExtraSuspension(), opAnn.blocking(), opAnn.causesBlocking(),
                     opAnn.promptCancellation());
-                actorGenerators.add(actorGenerator);
+                testStructureBuilder.actorGenerators.add(actorGenerator);
                 // Get list of groups and add this operation to specified ones
                 String opGroup = opAnn.group();
                 if (!opGroup.isEmpty()) {
-                    OperationGroup operationGroup = groupConfigs.get(opGroup);
+                    OperationGroup operationGroup = testStructureBuilder.groupConfigs.get(opGroup);
                     if (operationGroup == null)
                         throw new IllegalStateException("Operation group " + opGroup + " is not configured");
                     operationGroup.actors.add(actorGenerator);
                 }
                 String opNonParallelGroupName = opAnn.nonParallelGroup();
                 if (!opNonParallelGroupName.isEmpty()) { // is `nonParallelGroup` specified?
-                    groupConfigs.computeIfAbsent(opNonParallelGroupName, name -> new OperationGroup(name, true));
-                    groupConfigs.get(opNonParallelGroupName).actors.add(actorGenerator);
+                    testStructureBuilder.groupConfigs.computeIfAbsent(opNonParallelGroupName, name -> new OperationGroup(name, true));
+                    testStructureBuilder.groupConfigs.get(opNonParallelGroupName).actors.add(actorGenerator);
                 }
             }
             if (m.isAnnotationPresent(Validate.class)) {
                 if (m.getParameterCount() != 0)
                     throw new IllegalStateException("Validation function " + m.getName() + " should not have parameters");
-                validationFunctions.add(new Actor(m, Collections.emptyList()));
+                if (testStructureBuilder.validationFunction == null) {
+                    testStructureBuilder.validationFunction = new Actor(m, Collections.emptyList());
+                } else {
+                    throw new IllegalStateException("You can't have more than one validation function.\n" +
```

Comment:
Let's keep the logic that obtains all validation functions, check that the list size is <= 1, and print all the validation functions if several ones are listed.